src/Generator/MethodGenerator.php

```diff
@@ -378,12 +379,30 @@ private static function extractReturnTypeFromMethodReflection(MethodReflection $
             return null;
         }
 
-        $returnTypeString = (string) $returnType;
+        if (! method_exists($returnType, 'getName')) {
+            return self::expandLiteralType((string) $returnType, $methodReflection);
+        }
+
+        return ($returnType->allowsNull() ? '?' : '')
+            . self::expandLiteralType($returnType->getName(), $methodReflection);
+    }
 
-        if ('self' === strtolower($returnType)) {
+    /**
+     * @param string           $literalReturnType
+     * @param ReflectionMethod $methodReflection
+     *
+     * @return string
+     */
+    private static function expandLiteralType($literalReturnType, ReflectionMethod $methodReflection)
```

Comment:
Am I reading this correctly in that it pulls the type from reflecting the docblock in this case? If so, should it also consider `static` and `$this`? (If it's strictly from declared return type, ignore.)